src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/TextCompletionItemProvider.cs

```diff
@@ -0,0 +1,96 @@
+            {
+                throw new ArgumentNullException(nameof(htmlFactsService));
+            }
+
+            _htmlFactsService = htmlFactsService;
+        }
+
+        public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorSyntaxTree syntaxTree, TagHelperDocumentContext tagHelperDocumentContext, SourceSpan location)
+        {
+            var completions = new List<RazorCompletionItem>();
+            if (syntaxTree == null)
+            {
+                return completions;
+            }
+
+            var change = new SourceChange(location, string.Empty);
+            var owner = syntaxTree.Root.LocateOwner(change);
+
+            if (owner == null)
+            {
+                Debug.Fail("Owner should never be null.");
+                return completions;
+            }
+
+            if (AtTextCompletionPoint(owner))
+            {
+                var parent = owner.Parent;
+                if (_htmlFactsService.TryGetElementInfo(parent, out var containingTagNameToken, out var attributes) &&
+                    containingTagNameToken.Span.IntersectsWith(location.AbsoluteIndex))
+                {
+                    if (TextTag.StartsWith(containingTagNameToken.Content, StringComparison.OrdinalIgnoreCase)) {
+                        var completionDisplayText = TextTag;
+                        var completionItem = new RazorCompletionItem(
+                            completionDisplayText,
+                            completionDisplayText,
+                            RazorCompletionItemKind.Text,
+                            ElementCommitCharacters);
+                        var completionDescription = new DirectiveCompletionDescription("Create a inline HTML element without tags.");
+                        completionItem.SetDirectiveCompletionDescription(completionDescription);
+                        completions.Add(completionItem);
+                    }
+                }
+            }
+
+            return completions;
+        }
+
+        internal static bool AtTextCompletionPoint(Microsoft.AspNetCore.Razor.Language.Syntax.SyntaxNode owner)
+        {
+            // Only provide IntelliSense for C# code blocks, of the form:
+            // @{ }, @code{ }, @functions{ }, @if(true){ }
+            var implicitExpression = owner.FirstAncestorOrSelf<CSharpCodeBlockSyntax>();
+            if (implicitExpression == null)
+            {
+                return false;
+            }
+
+            if (implicitExpression.FirstAncestorOrSelf<MarkupElementSyntax>() != null)
+            {
+                // Implicit expression is nested in an HTML element
+                return false;
+            }
+
+            if (implicitExpression.FirstAncestorOrSelf<MarkupTagHelperElementSyntax>() != null)
+            {
+                // Implicit expression is nested in a TagHelper
+                return false;
+            }
+
+            return true;
+        }
```

Comment:
Does this logic work for the following cases?
```razor
@if (true)
{
<div>
  < @* Shouldn't get text completion here *@ 
</div>
}
```

```razor
<div>
@if (true)
{
  < @* Should get text completion here *@ 
}
</div>
```

I suspect they might not work.

Reply:
You might also get the completion in single line transitions:
```Razor
@{
    @: Here's some Markup | <-- You shouldn't get a <text> tag completion here.
}
```

Reply:
The latest commit (https://github.com/dotnet/aspnetcore-tooling/pull/1675/commits/e9b14990422da840f9f7552c8772266bb1d289ce) attempts to address this issue. Per @ajaybhargavb's suggestion, I walk up the syntax tree and try to determine the language in the given context (markup vs C#).

Reply:
Note, valid `<text>` nodes would always be in a `MarkupBlockSyntax`. Consequently, I examine the `MarkupElementSyntax`'s number of children. If there's only 1 child, that would indicate the `<text>` node is on its own. Ex.

```
<div>
@if (true)
{
  < @* Should get text completion here *@ 
}
</div>
```

While multiple children would indicate the form:
```
@if (true)
{
<div>. <--- CHILD 1
  < @* Shouldn't get text completion here *@ <--- CHILD 2
</div>
}
```

Reply:
Looks like the following case from the original issue still needs investigation:

```
@code {
    public void SomeFunctionAcceptingMethod(Func s) {

    }
}

@SomeFunctionAcceptingMethod(() =>
{
    <te <--- Text should autocomplete here
})


@(SomeFunctionAcceptingMethod(() =>
{
    <te <--- Text should autocomplete here
}))
```